File: src/ts/data-structures/Stack.ts
type Node<T> = {
    value: T;
    next?: Node<T>;
};
export default class Stack<T> {
    public length: number;
    private tail?: Node<T>;

    constructor() {
        this.tail = undefined;
        this.length = 0;
    }

    push(item: T): void {
        this.length++;
        const node: Node<T> = {
            value: item,
        };
        if (!this.tail) {
            this.tail = node;
        }
        node.next = this.tail;
        this.tail = node;
    }

    pop(): T | undefined {
        if (!this.tail) return;

        this.length--;
        const tail = this.tail;

        if (this.length === 0) {
            this.tail = undefined;
        } else {
            this.tail = tail.next;
        }

        return tail.value;
    }

    peek(): T | undefined {
        return this.tail?.value;
    }
}

File: src/data-structures/java/Stack.java
public class Node<T> {
  public T val;
  public Node<T> next;

  public Node(T val, Node<T> next) {
    this.val = val;
    this.next = next;
  }
  public Node() {
    this.val = null;
    this.next = null;
  }

}

public class Stack<T> {
  public int length; 
  private Node<T> tail;

  public Stack() {
    this.length = 0;
    this.tail = null;
  }

  public void Push(T item) {
    this.length++;
    Node<T> newNode = new Node<>();
    newNode.val = item;

    if (this.length == 1) {
      this.tail = newNode;
    } else {
      newNode.next = this.tail;
      this.tail = newNode;      
    }

  }

  public T Pop() {
    if(this.tail == null) return null;

    this.length--;

    Node<T> prevTail = this.tail;

    if (this.length == 0) {
      this.tail = null;
    } else {
      this.tail = this.tail.next;    
    }

    return prevTail.val;
  }

  public T Peek() {
    return this.tail != null ? this.tail.val : null;
  }
}